### PR TITLE
Pseudo likelihood

### PR DESCRIPTION
**Pull request by pamatteo**

### CODE REVIEW

File: src/inference/graph_inference.cpp [hunk 1]
```diff
@@ -341,6 +342,24 @@ class GraphNodeAssignment : public Nice2Assignment {
     return label_set_->GetLabelName(label_id);
   }
 
+  // Returns conditional probability of a label conditionated on its neighbouring labels
+  double GetNodeConditionalProbabilityGivenNeighbours(const GraphInference& fweights, int node, const double normalization, int label) const {
+    return exp(GetNodeScoreGivenAssignmentToANode(fweights, node, node, label)) / normalization;
+  }
+
+  // Returns the normalization related to a certain node
+  double GetNodeNormalizationConstantGivenNeighbours(const GraphInference& fweights, int node, std::vector<int>* candidates) const {
```
Comment (vraychev) [inline src/inference/graph_inference.cpp]: I see some vector being appended. This method does not only compute normalization.

File: src/inference/graph_inference.cpp [hunk 2]
```diff
@@ -1089,6 +1148,36 @@ void GraphInference::SSVMLearn(
     }
   }
 
+}
+
+void GraphInference::CommonInit(double regularization) {
```
Comment (vraychev) [inline src/inference/graph_inference.cpp]: rename to InitializeFeatureWeights

File: src/inference/graph_inference.cpp [hunk 3]
```diff
@@ -341,6 +342,24 @@ class GraphNodeAssignment : public Nice2Assignment {
     return label_set_->GetLabelName(label_id);
   }
 
+  // Returns conditional probability of a label conditionated on its neighbouring labels
```
Comment (vraychev) [inline src/inference/graph_inference.cpp]: I see no benefit of having these two methods. The formula would be far more understandable if you just include the calculations in PLLearn.

File: src/inference/graph_inference.cpp [hunk 4]
```diff
@@ -1101,10 +1190,46 @@ void GraphInference::SSVMLearn(
...
+    }
+  }
+}
+
+void GraphInference::PLLearn(
+    const Nice2Query* query,
+    const Nice2Assignment* assignment,
+    double learning_rate,
+    PrecisionStats* stats,
+    int num_pass) {
```
Comment (vraychev) [inline src/inference/graph_inference.cpp]: Unused

File: src/inference/graph_inference.cpp [hunk 5]
```diff
@@ -1101,10 +1190,46 @@ void GraphInference::SSVMLearn(
...
+      }
+    }
+  }
+}
+
+void GraphInference::PLLearn(
+    const Nice2Query* query,
+    const Nice2Assignment* assignment,
+    double learning_rate,
+    PrecisionStats* stats,
```
Comment (vraychev) [inline src/inference/graph_inference.cpp]: Unused. Remove.

File: src/inference/graph_inference.cpp [hunk 6]
```diff
@@ -537,6 +585,30 @@ class GraphNodeAssignment : public Nice2Assignment {
     }
   }
 
+  // Method that given a certain node and a label, first assign that label to the given node and then add the gradient_weight
+  // to every affected feature related to the given node and its neighbours
+  void GetNeghbouringAffectedFeature(
```
Comment (vraychev) [inline src/inference/graph_inference.cpp]: Neighboring (US)

and features are plural.

File: src/inference/graph_inference.cpp [hunk 7]
```diff
@@ -537,6 +585,30 @@ class GraphNodeAssignment : public Nice2Assignment {
...
+      GraphInference::SimpleFeaturesMap* affected_features,
+      int node,
+      int label,
+      double gradient_weight) const {
+    for (auto & arc : query_->arcs_adjacent_to_node_[node]) {
```
Comment (vraychev) [inline src/inference/graph_inference.cpp]: const auto&

File: src/training/train.cpp [hunk 8]
```diff
@@ -41,11 +44,22 @@ DEFINE_double(start_learning_rate, 0.1, "Initial learning rate");
 DEFINE_double(stop_learning_rate, 0.0001, "Stop learning if learning rate falls below the value");
 DEFINE_double(regularization_const, 2.0, "Regularization constant. The higher, the more regularization.");
 DEFINE_double(svm_margin, 0.1, "SVM Margin = Penalty for keeping equal labels as in the training data during training.");
+DEFINE_int32(beam_size, 5, "Beam size used to get the labels space for the normalization function when training with pseudolikelihood");
 
 DEFINE_int32(cross_validation_folds, 0, "If more than 1, cross-validation is performed with the specified number of folds");
 DEFINE_bool(print_confusion, false, "Print confusion statistics instead of training.");
 
 DEFINE_bool(train_multiclass_classifier, false, "Perform training by base multiclass classifier.");
+DEFINE_bool(train_pseudolikelihood, false, "Perform training on pseudolikelihood optimization.");
```
Comment (vraychev) [inline src/training/train.cpp]: Instead of 100 boolean flags, you may want to create a string flag with multiple values that you accept.

File: src/training/train.cpp [hunk 9]
```diff
@@ -157,27 +171,85 @@ void TestInference(RecordInput* input, GraphInference* inference) {
 
 }
 
-void Train(RecordInput* input, GraphInference* inference) {
-  inference->SSVMInit(FLAGS_regularization_const, FLAGS_svm_margin);
+void Train(RecordInput* input, GraphInference* inference, int fold_id) {
```
Comment (vraychev) [inline src/training/train.cpp]: This method is full of if statements about PL / SSVM. There is really not much benefit in having one method and not two for the two cases. It also makes it quite bad with the combination of PL+SSVM (if it would be useful in the end).

File: src/training/train.cpp [hunk 10]
```diff
@@ -157,27 +171,85 @@ void TestInference(RecordInput* input, GraphInference* inference) {
...
+      if (FLAGS_learning_prop_sqrt_pass == true) {
+        learning_rate /= pow(pass + 1, 0.5);
+      } else if (FLAGS_learning_prop_pass == true) {
+        learning_rate /= (pass + 1);
+      } else if (FLAGS_learning_prop_pass_and_initial_learn == true) {
+        learning_rate = initial_learning_rate / (1 + FLAGS_pl_lambda * (pass + 1));
+      }
+    }
+
+    if (FLAGS_profile) {
```
Comment (vraychev) [inline src/training/train.cpp]: I guess you can remove the profiling. Not everyone will download gprofiler in order to compile Nice2Predict. Don't forget to remove the include and also include the fold_id param.

File: src/inference/graph_inference.cpp (hunk 3 above)
Comment (pamatteo) [inline src/inference/graph_inference.cpp]: I disagree a little bit on this, since if I include these two methods especially the one computing the normalisation factor then the PLLearn will become longer and I think less readable. In addition, I think the computation of the marginals and the estimation of the normalisation factor belongs more on a separated function than inside the learning procedure. However, if you think it would be more readable moving all the code into the PLLearn function I will do it, so let me know.

Comment (vraychev) [inline src/inference/graph_inference.cpp]: Given that GetNodeNormalizationConstantGivenNeighbours has some "hidden" functionality with appending to the vector, I am not convinced it is a good idea to have the methods. You can have the methods if you can clearly describe what they do. The first on is just a formula that you call once, the second one you cannot describe even with the super long name.

Comment (pamatteo) [inline src/inference/graph_inference.cpp]: Now I understand which functions you were referring to (GetNodeConditionalProbabilityGivenNeighbours and GetNodeNormalizationConstantGivenNeighbours). They are small and I can move them inside PLLearn. I agree with you they are better inside PLLearn.

File: src/inference/graph_inference.cpp (hunk 1 above)
Comment (pamatteo) [inline src/inference/graph_inference.cpp]: In the normalisation, the first thing that I do is to add to the candidates vector the current  assignment of the node to which I am computing the estimated normalisation. As I said before now I see why this is better inside PLLearn. It would be more readable.

File: src/inference/graph_inference.cpp (hunk 6 above)
Comment (pamatteo) [inline src/inference/graph_inference.cpp]: Got it.

File: src/inference/graph_inference.cpp (hunk 5 above)
Comment (pamatteo) [inline src/inference/graph_inference.cpp]: In this comment I am a bit confused what are you referring to when you say unused?

File: src/training/train.cpp (hunk 9 above)
Comment (pamatteo) [inline src/training/train.cpp]: Are you saying that I should have one common initialisation that initialise all the various parameters, and not two different for the two trainings?

Comment (vraychev) [inline src/training/train.cpp]: It is not about the initialization, but about the loop in Train. This was a small method that grew substantially with a lot of ifs you add and these ifs are hard to read.
